### ISSUE TITLE
unconditional `post_exec`?

### ISSUE THREAD

**Issue by andre-merzky** [radical-cybertools/radical.entk#298]
I like the current `post_exec` mechanism on stages - but semantically its more a post conditional, right?  Or maybe I am missing something.  Could you clarify what the difference would be between

```py
def check():
    if random.randint() % 2: return True
    else                   : return False

def on_true():
    print 'wohoo!'

def on_false():
    raise ValueError('oops')

stage.post_exec = {'condition' : check,
                   'on_true'   : on_true, 
                   'on_false'  : on_false}
```

and

```py
def check():
    if random.randint() % 2: print `wohoo!'
    else                   : raise ValueError('oops')

def on_true():
    pass

def on_false()
    pass

stage.post_exec = {'condition' : check,
                   'on_true'   : on_true, 
                   'on_false'  : on_false}
```
?

If that is the same, I would like to be able to leave `on_true` and `on_false` unset - the current use (declaring dummy methods) is convoluting the meaning.

Maybe:
```py
stage.post_exec = check
```
would suffice?  But I am not sure how  EnTK  handles type differentiation like that in general...

**Comment by vivek-bala** [comment]
Hey Andre, I think you're on spot on. We can leave the entire decision making to one function. Also, 'on_true' and 'on_false' give the impression that decisions can only be binary. I wanted to remove it for that reason as well. I agree leaving it to 'check' would be the correct approach.

**Comment by andre-merzky** [comment]
Thanks for the clarification!  `post_exec` sounds like the proper name then really.  The function name should not matter anyway (outside of documentation and examples).

**Comment by vivek-bala** [comment]
I'll leave this open for now. Just so that I remember to make the fix during the release.

**Comment by vivek-bala** [comment]
PR #297 fixes this issue.

**Comment by vivek-bala** [comment]
Apologies on the incorrect reference. #297 does not address this ticket. This ticket will be changed in the next release as it is an API change.

**Comment by andre-merzky** [comment]
BTW, this change can be kept backward compatible:

```python

if isininstance(post_exec, dict):
  def _post_exec():
        if post_exec['condition'](): 
            post_exec['on_true']()
        else:
            post_exec['on_false']()
  post_exec = _post_exec
```
or something like that...  You could add a deprecation warning for a couple releases before removing it...

**Comment by vivek-bala** [comment]
Thanks Andre. That makes sense. I'll have an intermediate release before the next scheduled one with the deprecation. Both, for this ticket and #295 .

**Comment by vivek-bala** [comment]
We have moved to the post_exec being a callable function. We no longer accept a dictionary with 3 keys+values. Released in 0.7.14.
